Add Pix2Struct

younesbelkada opened the pull request:
# What does this PR do?

Fixes #20663

Paper: https://arxiv.org/pdf/2210.03347.pdf
Code: https://github.com/google-research/pix2struct

`Pix2Struct` is a series of Image-text models that has been fine-tuned on various datasets and tasks.
![Screenshot 2023-03-10 at 09 42 19](https://user-images.githubusercontent.com/49240599/224267136-389b8300-6497-4321-a6cc-d85b4f8e55d7.png)

This integration will offer users variety of models and potential use cases

`Pix2Struct` is a model that combines vision encoder and text decoder, similar as T5. The method heavily relies on its image processing procedure. The image pre-proccessing differs from classic Vision Transformers by being able to handle images of variable resolution, thus being able to keep the aspect ratio of the original image, that seems to be essential and crucial for image understanding. 

![Screenshot 2023-03-10 at 09 47 12](https://user-images.githubusercontent.com/49240599/224268204-f97d76b4-54e4-470a-aa44-9b6c086c8369.png)

Therefore I decided to change the current paradigm for getting `pixel_values` differently. Now the pixel values should be seen as tokens that are directly processed by the `ImageProcessor`. Hence, I decided to change `pixel_values` to `pixel_embeds` , as in fact they correspond to the image embeddings. We now obtain the patch embeddings directly from the processor, that is also responsible of also computing the pixel embeds attention mask.

I will update all the weights (18 in total) after I get 1 approval

### TODO

- FIne-tuning notebook

younesbelkada committed:
v1 all keys match

younesbelkada committed:
clean up

younesbelkada committed:
forward pass ok

younesbelkada committed:
add correct image transform

younesbelkada committed:
generate works, logits matching

younesbelkada committed:
clean up

younesbelkada committed:
more refactor

younesbelkada committed:
revert

younesbelkada committed:
revert

younesbelkada committed:
clean up

younesbelkada committed:
clean ups

younesbelkada committed:
clean up

younesbelkada committed:
refactor

younesbelkada committed:
refactor

younesbelkada committed:
fix doc

younesbelkada committed:
fix tokenizer test

younesbelkada committed:
fix toctree

younesbelkada committed:
revert toctree

younesbelkada committed:
oops

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21400). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
few fixes

younesbelkada committed:
replace to `pixel_embeds`

younesbelkada committed:
make fixup

younesbelkada committed:
test processing & feat extractor

younesbelkada committed:
Merge branch 'main' into add-pix2struct-clean

younesbelkada committed:
fix some tests

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
more fixes

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
make fixup

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
clean up

younesbelkada committed:
more clean up

younesbelkada committed:
add a single slow test

younesbelkada committed:
fix test

younesbelkada committed:
make fixup

younesbelkada committed:
fix

ArthurZucker commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,441 @@
+                "'gated-gelu' or 'relu'"
+            )
+
+        # for backwards compatibility
+        if feed_forward_proj == "gated-gelu":
+            self.dense_act_fn = "gelu_new"
```
we probably don'tneed this

younesbelkada replied:
fixed

ArthurZucker commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,441 @@
+        num_hidden_layers (`int`, *optional*, defaults to 12):
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        image_size (`int`, *optional*, defaults to 224):
+            The size (resolution) of each image.
+        patch_size (`int`, *optional*, defaults to 32):
+            The size (resolution) of each patch.
+        dense_act_fn (`str` or `function`, *optional*, defaults to `"gelu"`):
+            The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+            `"relu"`, `"selu"` and `"gelu_new"` ``"gelu"` are supported. layer_norm_eps (`float`, *optional*, defaults
```
some args are missing

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+ALL_LAYERNORM_LAYERS.append(Pix2StructLayerNorm)
+
+
+class Pix2StructVisionEmbeddings(nn.Module):
```
is this copied from?

younesbelkada replied:
Nope!

younesbelkada replied:
Will update the docstring

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        self.dropout = config.attention_dropout
+        self.inner_dim = self.n_heads * self.key_value_proj_dim
+
+        # Mesh TensorFlow initialization to avoid scaling before softmax
+        self.query = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.key = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.value = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.o = nn.Linear(self.inner_dim, self.hidden_size, bias=False)
```
if this is not copied from, better variable namesplese (o_

younesbelkada replied:
Fixed!

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        value_states = shape(self.value(hidden_states))
+
+        # compute scores
+        scores = torch.matmul(
+            query_states, key_states.transpose(3, 2)
+        )  # equivalent of torch.einsum("bnqd,bnkd->bnqk", query_states, key_states), compatible with onnx op>9
```
com can be placed on top

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        position_bias_masked = position_bias
+        if position_bias_masked is not None:
+            # replace 0 by -inf
+            position_bias_masked = position_bias_masked.masked_fill(position_bias_masked == 0, -10000.0)
+            position_bias_masked = position_bias_masked.masked_fill(position_bias_masked != -10000, 0)
```
-1000 hard coded can it not be the case?

younesbelkada replied:
Fixed!

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+
+        scores += position_bias_masked
+
+        attn_weights = nn.functional.softmax(scores.float(), dim=-1).type_as(
+            scores
+        )  # (batch_size, n_heads, seq_length, key_length)
```
place comment above pls

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        decoder_start_token_id = self.config.decoder_start_token_id
+        pad_token_id = self.config.pad_token_id
+
+        assert decoder_start_token_id is not None, (
+            "self.model.config.decoder_start_token_id has to be defined. In Pix2StructText it is usually set to the"
+            " pad_token_id. See Pix2StructText docs for more information"
+        )
+
```
no assert unless this is copied from

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        is_small = relative_position < max_exact
+
+        # The other half of the buckets are for logarithmically bigger bins in positions up to max_distance
+        relative_position_if_large = max_exact + (
+            torch.log(relative_position.float() / max_exact)
+            / math.log(max_distance / max_exact)
+            * (num_buckets - max_exact)
+        ).to(torch.long)
```
split in two lines

younesbelkada replied:
Actually this is a copied from

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        real_seq_length = seq_length
+
+        if past_key_value is not None:
+            assert (
```
no asserts

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+        else:
+            outputs = outputs + attention_outputs
+
+        return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
```
to remove ?

ArthurZucker commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1686 @@
+                    layer_past_state.index_select(0, beam_idx.to(layer_past_state.device)),
+                )
+
+            assert reordered_layer_past_states[0].shape == layer_past_states[0].shape
+            assert len(reordered_layer_past_states) == len(layer_past_states)
+
```
no assert unless it is a legacy related to a copied from

ArthurZucker reviewed (commented):
thanks for this! Missing a bunch of copied from it seems no?

ankrgyl commented:
@younesbelkada @ArthurZucker 👋 how is this PR going? Do you need some help to get it over the finish line? Happy to collab if helpful.

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
fix authors

younesbelkada commented on README.md:
```diff
@@ -388,6 +388,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[PEGASUS-X](https://huggingface.co/docs/transformers/model_doc/pegasus_x)** (from Google) released with the paper [Investigating Efficiently Extending Transformers for Long Input Summarization](https://arxiv.org/abs/2208.04347) by Jason Phang, Yao Zhao, and Peter J. Liu.
 1. **[Perceiver IO](https://huggingface.co/docs/transformers/model_doc/perceiver)** (from Deepmind) released with the paper [Perceiver IO: A General Architecture for Structured Inputs & Outputs](https://arxiv.org/abs/2107.14795) by Andrew Jaegle, Sebastian Borgeaud, Jean-Baptiste Alayrac, Carl Doersch, Catalin Ionescu, David Ding, Skanda Koppula, Daniel Zoran, Andrew Brock, Evan Shelhamer, Olivie ...
 1. **[PhoBERT](https://huggingface.co/docs/transformers/model_doc/phobert)** (from VinAI Research) released with the paper [PhoBERT: Pre-trained language models for Vietnamese](https://www.aclweb.org/anthology/2020.findings-emnlp.92/) by Dat Quoc Nguyen and Anh Tuan Nguyen.
+1. **[Pix2Struct](https://huggingface.co/docs/transformers/main/model_doc/pix2struct)** (from <FILL INSTITUTION>) released with the paper [<FILL PAPER TITLE>](<FILL ARKIV LINK>) by <FILL AUTHORS>.
```
Fixed in [c95b045](https://github.com/huggingface/transformers/pull/21400/commits/c95b0458f1394cd9558641dd267ac4d0ab8b9055)

younesbelkada committed:
fix toctree

younesbelkada committed:
update docs

younesbelkada committed:
add docstring

younesbelkada committed:
revert change

younesbelkada committed:
Update src/transformers/models/pix2struct/__init__.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
fix tokenizer

younesbelkada commented on src/transformers/models/pix2struct/processing_pix2struct.py:
```diff
@@ -0,0 +1,156 @@
+    """
+    attributes = ["image_processor", "tokenizer"]
+    image_processor_class = "Pix2StructImageProcessor"
+    tokenizer_class = ("T5Tokenizer", "T5TokenizerFast")
```
Fixed!

younesbelkada committed:
fix processor test

younesbelkada commented on tests/models/pix2struct/test_image_processing_pix2struct.py:
```diff
@@ -0,0 +1,266 @@
+        or a list of PyTorch tensors if one specifies torchify=True.
+        """
+
+        assert not (numpify and torchify), "You cannot specify both numpy and PyTorch tensors at the same time"
```
Fixed!

younesbelkada committed:
fix test

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
make fixup

younesbelkada committed:
refactor

younesbelkada committed:
fix config

younesbelkada committed:
Update src/transformers/models/pix2struct/image_processing_pix2struct.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
format

younesbelkada committed:
fix

younesbelkada committed:
Update src/transformers/models/pix2struct/image_processing_pix2struct.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

younesbelkada committed:
make fixup

younesbelkada committed:
add docstring

younesbelkada committed:
fix issues

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
add slow test

younesbelkada committed:
fix

younesbelkada committed:
fix

younesbelkada committed:
fix batched issue

younesbelkada committed:
fix training issues

younesbelkada committed:
fix ci test

younesbelkada committed:
fix slow test

ankrgyl commented:
It looks like you've got it under control so I'll bow out, but happy to test!

younesbelkada committed:
fix conversion script

younesbelkada commented on src/transformers/models/pix2struct/convert_pix2struct_original_pytorch_to_hf.py:
```diff
@@ -0,0 +1,193 @@
+
+    image_size = 384
+    image = load_demo_image(image_size=image_size, device="cpu")
+    tokenizer = Pix2StructTokenizer.from_pretrained("bert-base-uncased")
```
Updated the correct conversion script in [9dc2b25](https://github.com/huggingface/transformers/pull/21400/commits/9dc2b25a6e3f4a006428a92f1ede1aaad89ba9f5)

younesbelkada commented on src/transformers/models/pix2struct/convert_pix2struct_original_pytorch_to_hf.py:
```diff
@@ -0,0 +1,193 @@
+    out = hf_itm_model(question_input_ids, image, use_itm_head=False)
+
+    assert out[0].item() == 0.2110687494277954
+    assert torch.nn.functional.softmax(out_itm[0], dim=1)[:, 1].item() == 0.45698845386505127
```
Updated the correct conversion script in [9dc2b25](https://github.com/huggingface/transformers/pull/21400/commits/9dc2b25a6e3f4a006428a92f1ede1aaad89ba9f5)

younesbelkada committed:
remove unneeded classes

younesbelkada committed:
fix slow test

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,255 @@
+            image (`np.ndarray`):
+                Image to extract flattened patches from.
+        """
+        requires_backends(self.extract_flattened_patches, "torch")
+        # import torch
```
I am not sure as `@requires_torch` is only used for tests no?

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,255 @@
+
+        return result.float()
+
+    def normalize(
```
Here it's an image-wise normalization. We retrieve the mean and std of the input image and call `normalize` from image utils

younesbelkada replied:
```python
mean = np.mean(image)
std = np.std(image)
adjusted_stddev = max(std, 1.0 / math.sqrt(np.prod(image.shape)))

return normalize(image, mean=mean, std=adjusted_stddev, data_format=data_format, **kwargs)
```

younesbelkada committed:
fix require backends

younesbelkada committed:
fix masked fill

younesbelkada committed:
revert

younesbelkada committed:
fix softmax

younesbelkada committed:
add large models support

younesbelkada commented on src/transformers/utils/dummy_vision_objects.py:
```diff
@@ -387,6 +387,13 @@ def __init__(self, *args, **kwargs):
+class Pix2StructImageProcessor(metaclass=DummyObject):
+    _backends = ["vision"]
+
+    def __init__(self, *args, **kwargs):
```
Hum this makes `make fixup` fail

younesbelkada committed:
fix conditional generation

younesbelkada committed:
few fixes

younesbelkada committed:
add instructions

younesbelkada committed:
rm unneeded file

younesbelkada commented on src/transformers/models/pix2struct/convert_pix2struct_original_pytorch_to_hf.py:
```diff
@@ -0,0 +1,163 @@
+    raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB")
+
+    inputs = processor(raw_image, return_tensors="pt", max_patches=2048)
+    # inputs["decoder_input_ids"] = torch.LongTensor([[0]])
+
+    output = model.generate(**inputs, do_sample=False)
+
+    # assert processor.decode(output[0], skip_special_tokens=True) == "A stop sign is on a street corner."
+    print(processor.decode(output[0], skip_special_tokens=True))
```
```suggestion
    output = model.generate(**inputs, do_sample=False)

    print(processor.decode(output[0], skip_special_tokens=True))
```

younesbelkada committed:
Update src/transformers/models/pix2struct/convert_pix2struct_original_pytorch_to_hf.py

younesbelkada commented:
I think I have addressed most of the comments! 
I also updated the PR description and would love to have a round of review! 
cc @amyeroberts @ArthurZucker

younesbelkada committed:
fix ci test

younesbelkada committed:
fix ci test really

amyeroberts commented on tests/models/pix2struct/test_modeling_pix2struct.py:
```diff
@@ -0,0 +1,669 @@
+
+        self.assertEqual(
+            processor.decode(predictions[1], skip_special_tokens=True),
+            "A row of books including The Temple Bar and Guiness.",
```
📚 🍻

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        >>> decoder_input_ids = tokenizer("Studies show that", return_tensors="pt").input_ids  # Batch size 1
+
+        >>> # preprocess: Prepend decoder_input_ids with start token which is pad token for Pix2StructTextModel.
+        >>> # This is not needed for torch's Pix2StructTextForConditionalGeneration as it does this internally using labels arg.
```
Should this doc example be used for `Pix2StructTextModel` and a different one used for `Pix2StructTextForConditionalGeneration` ?

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+
+
+@add_start_docstrings(
+    "The bare Pix2StructText Model transformer outputting raw hidden-states without any specific head on top.",
```
copy-pasta: should be updated for `Pix2StructForConditionalGeneration`

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        labels=None,
+        return_dict=None,
+        **kwargs,
+    ):
```
Missing docstring

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        return outputs
+
+
+class Pix2StructVisionMlp(nn.Module):
```
This looks the same as `T5DenseGatedActDense` - if so can you add a `#copied from`?

younesbelkada replied:
Fixed in [47df080](https://github.com/huggingface/transformers/pull/21400/commits/47df080026bd945da876839efa1ae323e2fd11e4)

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        values = values.permute([2, 0, 1]).unsqueeze(0)  # shape (1, num_heads, query_length, key_length)
+        return values
+
+    def forward(
```
Large method, so hard to tell but this looks the same as the attention in T5.

younesbelkada replied:
Actually this is not entirely copied from T5 as we changed the logic in the forward pass a bit

amyeroberts reviewed (commented):
Thanks for adding this model! 

In general, PR is looking good and nothing major is missing :) 

Main comments: 
* Quite a few of the docstrings don't match the method/function signatures - including both the argument names and their default values
* I don't think `pixel_embeds` is the right name for the image processor output - these aren't embeddings but rather flattened patches with their ids appended. I would instead use something like `patches`
* Quite a few of the patterns in the image processor deviate from the rest of the library. In particular, all the preprocessing methods should accept and output numpy arrays; and how the behaviour is configured should be fully configurable from the function arguments. There's also some replication of logic in image_utils e.g. checking the channel dimension, switching the channel dimension format and getting the image dimensions. I've left comments to try and explain at relevant places - let me know if these aren't clear.

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
fix nit

younesbelkada committed:
fix nits

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+    ):
+        batch_size, _, _ = pixel_embeds.shape
+
+        vision_outputs = self.encoder(
```
Fixed in [f7e17a7](https://github.com/huggingface/transformers/pull/21400/commits/f7e17a747b24bde396b2511ca78bfac79e96fa83)

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        }
+
+
+Pix2StructText_START_DOCSTRING = r"""
```
Fixed in [f7e17a7](https://github.com/huggingface/transformers/pull/21400/commits/f7e17a747b24bde396b2511ca78bfac79e96fa83)

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+            configuration. Check out the [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+Pix2StructText_INPUTS_DOCSTRING = r"""
```
Fixed in [f7e17a7](https://github.com/huggingface/transformers/pull/21400/commits/f7e17a747b24bde396b2511ca78bfac79e96fa83) &
 3bad3b2

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+Pix2StructText_ENCODER_INPUTS_DOCSTRING = r"""
```
Fixed in 3bad3b2

younesbelkada committed:
fix image processors nits

younesbelkada commented on src/transformers/image_transforms.py:
```diff
@@ -719,3 +719,29 @@ def convert_to_rgb(image: ImageInput) -> ImageInput:
+
+
+# adapted from: https://discuss.pytorch.org/t/tf-image-extract-patches-in-pytorch/171409/2
+def torch_extract_patches(image_tensor, patch_height, patch_width):
```
Makes sense! Moved it in [f821d22](https://github.com/huggingface/transformers/pull/21400/commits/f821d2233682d1a95f6664ae4f9f8d4ff2f75a94)

younesbelkada committed:
docstring

younesbelkada commented on src/transformers/image_transforms.py:
```diff
@@ -719,3 +719,29 @@ def convert_to_rgb(image: ImageInput) -> ImageInput:
+# adapted from: https://discuss.pytorch.org/t/tf-image-extract-patches-in-pytorch/171409/2
+def torch_extract_patches(image_tensor, patch_height, patch_width):
+    """
+    Utiliy function to extract patches from a given image tensor.
```
Added more details in 4f713ef

younesbelkada committed:
clean up

younesbelkada commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,441 @@
+
+
+    Args:
+        hidden_size (`int`, *optional*, defaults to 768):
```
Fixed in [fb6f283](https://github.com/huggingface/transformers/pull/21400/commits/fb6f283e0b3349f596a8fe032d3aa2d3ead8343d)

younesbelkada committed:
fix nit

younesbelkada committed:
fix tests

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+    Constructs a PIX2STRUCT image processor.
+
+    Args:
+        do_resize (`bool`, *optional*, defaults to `True`):
```
Fixed in [d8d56c6](https://github.com/huggingface/transformers/pull/21400/commits/d8d56c65dfaf1563483871d97af3972435f26859)

younesbelkada commented on tests/models/pix2struct/test_image_processing_pix2struct.py:
```diff
@@ -0,0 +1,267 @@
+        self.patch_size = patch_size
+
+    def prepare_image_processor_dict(self):
+        return {
```
Fixed in 718027087ac8ee076333cd62bddd42c67d98718b

younesbelkada committed:
docstring nit

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        return shifted_input_ids
+
+
+VIT_START_DOCSTRING = r"""
```
Fixed in 2d8a1dc

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+            configuration. Check out the [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+VIT_INPUTS_DOCSTRING = r"""
```
Fixed in 2d8a1dc

younesbelkada committed:
fix reshape

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+            return states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
+
+        def unshape(states):
+            """reshape"""
```
Agreed! Fixed in [1603770](https://github.com/huggingface/transformers/pull/21400/commits/1603770b575409398078badfa4485d9805f21c52)

NielsRogge commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,267 @@
+
+class Pix2StructImageProcessor(BaseImageProcessor):
+    r"""
+    Constructs a PIX2STRUCT image processor.
```
```suggestion
    Constructs a Pix2Struct image processor.
```

NielsRogge commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,267 @@
+    Constructs a PIX2STRUCT image processor.
+
+    Args:
+        do_resize (`bool`, *optional*, defaults to `True`):
```
This argument doesn't seem to be used at init

younesbelkada replied:
Fixed in 8d5573f

younesbelkada committed:
Update src/transformers/models/pix2struct/image_processing_pix2struct.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

younesbelkada committed:
fix nit

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+
+VIT_INPUTS_DOCSTRING = r"""
+    Args:
+        pixel_embeds (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
```
Fixed in [7b912f7](https://github.com/huggingface/transformers/pull/21400/commits/7b912f74ef31412fab46b7feb92be9a9fee13551)

younesbelkada committed:
fix repetition

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
+        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
+
+        if attention_mask is None:
```
Fixed in [93695cd](https://github.com/huggingface/transformers/pull/21400/commits/93695cd1fe1cf2fd8c9c7af2cb912b29e85ca641)

younesbelkada committed:
refactor processor

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+        attention_masks = [(image.sum(dim=-1) != 0).float() for image in images]
+
+        # All transformations expect numpy arrays.
+        images = [to_numpy_array(image) for image in images]
```
Fixed in 213a69f !

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+            image (`np.ndarray`):
+                Image to extract flattened patches from.
+        """
+        requires_backends(self.extract_flattened_patches, "torch")
+
+        # convert to torch if
+        if not isinstance(image, torch.Tensor):
+            # check if channels first:
+            if image.shape[0] != 3:
+                image = torch.from_numpy(image).permute(2, 0, 1)
+            else:
+                image = torch.from_numpy(image)
+
+        patch_height, patch_width = self.patch_size
+        _, image_height, image_width = image.shape
+        image_height = float(image_height)
+        image_width = float(image_width)
+
+        # maximize scale s.t.
+        scale = math.sqrt(max_patches * (patch_height / image_height) * (patch_width / image_width))
+        num_feasible_rows = max(min(math.floor(scale * image_height / patch_height), max_patches), 1)
+        num_feasible_cols = max(min(math.floor(scale * image_width / patch_width), max_patches), 1)
+        resized_height = max(num_feasible_rows * patch_height, 1)
+        resized_width = max(num_feasible_cols * patch_width, 1)
+
+        image = torch.nn.functional.interpolate(
+            image.unsqueeze(0),
+            size=(resized_height, resized_width),
+            mode="bilinear",
+            align_corners=False,
+            antialias=True,
+        ).squeeze(0)
```
Thanks! 
This seemed to introduce some rounding errors (thus slightly changing the result of `generate` in a slow test) because we need to normalize before resizing as `pillow` does not support resizing on float tensors :/ I left the commit in another branch:  43324a84ee65caf60b7a6acb0b4d33247e720cd9 , will let it like this for now and dig further

younesbelkada committed:
make patch size consistent

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+    model_input_names = ["pixel_values"]
+
+    def __init__(
+        self, patch_size: Dict[str, int] = [16, 16], do_normalize: bool = True, do_convert_rgb: bool = True, **kwargs
```
Should be fixed in [efc492b](https://github.com/huggingface/transformers/pull/21400/commits/efc492b4ecb21218e1c1de19ba95d5c40ee4a705)

younesbelkada committed:
refactor forward

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+            else:
+                image = torch.from_numpy(image)
+
+        patch_height, patch_width = self.patch_size
```
Fixed in 48b8b6e

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+
+        return normalize(image, mean=mean, std=adjusted_stddev, data_format=data_format, **kwargs)
+
+    def preprocess(
```
Fixed in 48b8b6e

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+
+        Args:
+            image (`np.ndarray`):
+                Image to extract flattened patches from.
```
Fixed!

younesbelkada committed:
fix docstring

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+        Preprocess an image or batch of images.
+
+        Args:
+            images (`ImageInput`):
```
Fixed in [81f9ada](https://github.com/huggingface/transformers/pull/21400/commits/81f9ada18437f9b8cdbdd07c7339d71524a527e0)

younesbelkada committed:
fix max_patches issue

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+    def preprocess(
+        self,
+        images: ImageInput,
+        max_patches: Optional[int] = 2048,
```
Thanks! Fixed in [b3af4a5](https://github.com/huggingface/transformers/pull/21400/commits/b3af4a5b8cc1af4c56048237053f481f8fb6a735)

younesbelkada committed:
update docstirng

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,239 @@
+        do_normalize: Optional[bool] = None,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        do_convert_rgb: bool = None,
+        data_format: ChannelDimension = ChannelDimension.FIRST,
```
Hmm this argument seems to be needed after all by the `normalize` function. I have updated the docstring accordingly in [732dd25](https://github.com/huggingface/transformers/pull/21400/commits/732dd25156d5610f7d4b304cad710f9f44cd618e)

younesbelkada committed:
update docstring

younesbelkada committed:
fix coped from

younesbelkada commented on tests/models/pix2struct/test_image_processing_pix2struct.py:
```diff
@@ -0,0 +1,267 @@
+        raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB")
+        return raw_image
+
+    def prepare_inputs(self, equal_resolution=False, numpify=False, torchify=False):
```
Fixed in [f498734](https://github.com/huggingface/transformers/pull/21400/commits/f498734696d9d3f58f9496387fb22185d2d5076c)

younesbelkada committed:
add skip reasons

younesbelkada commented on tests/models/pix2struct/test_modeling_pix2struct.py:
```diff
@@ -0,0 +1,669 @@
+        pass
+
+    def test_retain_grad_hidden_states_attentions(self):
+        pass
```
Fixed in [38c0a76](https://github.com/huggingface/transformers/pull/21400/commits/38c0a766c0886821aebf601f7a99079ce143a226)

younesbelkada commented on tests/models/pix2struct/test_modeling_pix2struct.py:
```diff
@@ -0,0 +1,669 @@
+        pass
+
+    def test_training_gradient_checkpointing(self):
+        pass
```
Fixed in [38c0a76](https://github.com/huggingface/transformers/pull/21400/commits/38c0a766c0886821aebf601f7a99079ce143a226)

NielsRogge commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,282 @@
+        patch_size (`Dict[str, int]`, *optional*, defaults to `{"height": 16, "width": 16}`):
+            The patch size to use for the image. According to Pix2Struct paper and code, the patch size is 16x16.
+        do_normalize (`bool`, *optional*, defaults to `True`):
+            Whether to normalize the image. Can be overridden by the `do_normalize` parameter in the `preprocess`
+            method. Can be overridden by the `do_normalize` parameter in the `preprocess` method. According to
+            Pix2Struct paper and code, the image is normalized with its own mean and standard deviation.
```
```suggestion
            Whether to normalize the image. Can be overridden by the `do_normalize` parameter in the `preprocess`
            method. According to Pix2Struct paper and code, the image is normalized with its own mean and standard 
            deviation.
```

NielsRogge commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,282 @@
+            paper](https://arxiv.org/pdf/2210.03347.pdf).
+    """
+
+    model_input_names = ["pixel_values"]
```
This is probably going to be pixel_embeds?

younesbelkada committed:
few fixes

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        return outputs
+
+
+class Pix2StructTextLayerSelfAttention(nn.Module):
```
Fixed in [47df080](https://github.com/huggingface/transformers/pull/21400/commits/47df080026bd945da876839efa1ae323e2fd11e4)

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1698 @@
+        return outputs
+
+
+class Pix2StructTextLayerCrossAttention(nn.Module):
```
Fixed in [47df080](https://github.com/huggingface/transformers/pull/21400/commits/47df080026bd945da876839efa1ae323e2fd11e4)

younesbelkada committed:
Update src/transformers/models/pix2struct/image_processing_pix2struct.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

younesbelkada committed:
format

younesbelkada committed:
fix doctests

younesbelkada committed:
refactor and fix

younesbelkada committed:
fix doc build issue

younesbelkada commented:
Thanks @amyeroberts for the extensive review! Should have addressed most of them and left some open questions
Regarding the new naming `patches`, I am not 100% convinced about that, users needs to see these input as a new paradigm that is equivalent to text tokens (as there are also attention masks in this new input), and I am afraid `patches` will confuse users as the shape of this input would be hard to interpret `bs x seq_len x hidden_dim` (with `hidden_dim=num_channels x patch_width x patch_height`.)

younesbelkada committed:
fix processor test

younesbelkada committed:
small fix conversion script

younesbelkada committed:
replace correct weights

younesbelkada committed:
make fixup

younesbelkada committed:
fix some issues

amyeroberts reviewed (commented):
Nice work! 🤩  Only a few last things before we're good to merge

Notes: 
* The way of creating `text_config` and `vision_config` for CLIP-like configurations has been updated. There's a few bits of the configuration logic that should be updated to reflect that - I've left a more detailed note in `configuration_pix2struct.py`. As well as the code, we should also make sure that `text_config` and `vision_config` aren't uploaded in the config file on the hub. 
* There's a few remaining docstrings that should be updated so they reflect the functions

Re naming of the vision inputs - lets brainstorm to find something we both think works. I see your point about `patches` and how it could be confusing. Equally I think `pixel_embeds` is confusing as the embeddings are actually found in the `Pix2StructVisionEmbeddings` layer. I think the analogy to `input_ids` from text is a good one - is there something similar we could use here? e.g. `vision_input_ids`?

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,431 @@
+    ):
+        super().__init__(**kwargs)
+
+        # If `_config_dict` exist, we use them for the backward compatibility.
+        text_config_dict = kwargs.pop("text_config_dict", None)
+        vision_config_dict = kwargs.pop("vision_config_dict", None)
+        if text_config_dict is not None:
+            text_config = text_config_dict
+        if vision_config_dict is not None:
+            vision_config = vision_config_dict
```
Yeah you're right, will revert

younesbelkada committed:
revert config and fixes

younesbelkada commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,431 @@
+        if not isinstance(vision_config, Pix2StructVisionConfig):
+            self.vision_config = Pix2StructVisionConfig(**vision_config)
+        else:
+            self.vision_config = vision_config
```
Fixed!

younesbelkada commented on src/transformers/models/pix2struct/configuration_pix2struct.py:
```diff
@@ -0,0 +1,431 @@
+        dropout_rate (`float`, *optional*, defaults to 0.0):
+            The dropout probabilitiy for all fully connected layers in the embeddings, encoder, and pooler.
+        hidden_dropout_prob (`float`, *optional*, defaults to 0.0):
+            The dropout ratio for the attention probabilities.
```
Fixed!

younesbelkada committed:
Update src/transformers/models/pix2struct/image_processing_pix2struct.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
more details

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,277 @@
+        self, image: np.ndarray, data_format: Optional[Union[str, ChannelDimension]] = None, **kwargs
+    ) -> np.ndarray:
+        """
+        Normalize an image. image = (image - image_mean) / image_std.
```
Done!

younesbelkada committed:
fixes

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,277 @@
+        Preprocess an image or batch of images.
+
+        Args:
+            images (`ImageInput`):
```
Fixed!

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,277 @@
+        **kwargs,
+    ) -> PIL.Image.Image:
+        """
+        Preprocess an image or batch of images.
```
Sure, fixed!

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+        >>> # forward pass
+        >>> outputs = model.generate(**inputs)
+        >>> print(processor.batch_decode(outputs, skip_special_tokens=True))
+        ['A stop sign the street with a sign that says yes']
```
Yeah I think the model struggles when it comes to guided generation

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,277 @@
+        # convert to torch if
+        if not isinstance(image, torch.Tensor):
+            image = to_channel_dimension_format(image, ChannelDimension.FIRST)
+            image = torch.from_numpy(image)
```
Thanks a lot for the explanation!

younesbelkada committed:
fix processor

younesbelkada committed:
fix processor test

younesbelkada commented on tests/models/pix2struct/test_image_processing_pix2struct.py:
```diff
@@ -0,0 +1,261 @@
+        raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB")
+        return raw_image
+
+    # Copied from transformers.tests.test_image_processing_common.prepare_image_inputs
```
Fixed!

younesbelkada commented on tests/models/pix2struct/test_image_processing_pix2struct.py:
```diff
@@ -0,0 +1,261 @@
+        size=None,
+        do_normalize=True,
+        do_convert_rgb=True,
+        patch_size=16,
```
Fixed!

younesbelkada committed:
fix

younesbelkada commented on tests/models/pix2struct/test_modeling_pix2struct.py:
```diff
@@ -0,0 +1,674 @@
+            self.assertDictEqual(config.text_config.to_dict(), text_config.to_dict())
+
+
+# We will verify our results on an image of cute cats
```
Adapted the comment !

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
make fixup

NielsRogge commented on tests/models/pix2struct/test_modeling_pix2struct.py:
```diff
@@ -593,7 +593,7 @@ def test_load_vision_text_config(self):
-# We will verify our results on an image of cute cats
+# We will verify our results on an image of a stop sign
 def prepare_img():
     url = "https://www.ilankelman.org/stopsigns/australia.jpg"
```
Please don't use a URL from another website in the tests, better to host on HF

younesbelkada replied:
Fixed, thanks for the heads up

younesbelkada committed:
fix processor

younesbelkada commented on src/transformers/models/pix2struct/processing_pix2struct.py:
```diff
@@ -0,0 +1,158 @@
+    def __init__(self, image_processor, tokenizer):
+        tokenizer.return_token_type_ids = False
+        super().__init__(image_processor, tokenizer)
+        self.current_processor = self.image_processor
```
Fixed!

younesbelkada committed:
Update src/transformers/models/pix2struct/modeling_pix2struct.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+        self.wi_0 = nn.Linear(config.hidden_size, config.d_ff, bias=False)
+        self.wi_1 = nn.Linear(config.hidden_size, config.d_ff, bias=False)
+        self.wo = nn.Linear(config.d_ff, config.hidden_size, bias=False)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Fixed!

younesbelkada committed:
add copied

younesbelkada committed:
Merge branch 'add-pix2struct-clean' of https://github.com/younesbelkada/transformers into add-pix2struct-clean

younesbelkada committed:
make fixup

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+        )
+
+
+class Pix2StructTextDenseGatedActDense(nn.Module):
```
Fixed!

younesbelkada committed:
fix copies

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+        self.layer_norm = Pix2StructLayerNorm(config.hidden_size, eps=config.layer_norm_epsilon)
+        self.dropout = nn.Dropout(config.dropout_rate)
+
+    def forward(self, hidden_states):
```
Fixed!

younesbelkada committed:
update docstring

younesbelkada commented on tests/models/pix2struct/test_processor_pix2struct.py:
```diff
@@ -0,0 +1,192 @@
+        shutil.rmtree(self.tmpdirname)
+
+    def prepare_image_inputs(self):
+        """This function prepares a list of PIL images, or a list of numpy arrays if one specifies numpify=True,
```
I updated the docstring instead

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+        elif return_dict and not isinstance(encoder_outputs, BaseModelOutput):
+            encoder_outputs = BaseModelOutput(
+                last_hidden_state=encoder_outputs[0],
+                hidden_states=encoder_outputs[1] if len(encoder_outputs) > 1 else None,
+                attentions=encoder_outputs[2] if len(encoder_outputs) > 2 else None,
```
Actually we never enter in this case no? if `return_dict` then the encoder will always return `BaseModelOutput`. Will remove this block

younesbelkada replied:
Fixed in [98c02dc](https://github.com/huggingface/transformers/pull/21400/commits/98c02dca5bb1bb5959ba94875b509df2dfa74156)

younesbelkada committed:
refactor

younesbelkada committed:
fix docstring

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1722 @@
+    def get_encoder(self):
+        return self.encoder
+
+    @add_start_docstrings_to_model_forward(PIX2STRUCT_INPUTS_DOCSTRING)
```
Thanks for the heads up! Fixed in [a8b7a47](https://github.com/huggingface/transformers/pull/21400/commits/a8b7a478f634f57cb9a21e71e6b8b3150b1ad908)

younesbelkada committed:
fix conversion script

younesbelkada committed:
fix vqa issue

younesbelkada committed:
replace to `flattened_patches`

younesbelkada committed:
nit

younesbelkada committed:
fix numpy issue

younesbelkada commented:
As disccused offline, let's stick for `flattened_patches` ! I should have fixed your comments by now and added support for `vqa` models in `Pix2struct` as they require a specific format / way of inferring

amyeroberts commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -282,6 +282,17 @@ def render_header(self, image: ImageInput, header: str, **kwargs):
+        return_framework = None
+
+        # Convert to PIL image if necessary
+        if not isinstance(image, Image.Image):
```
There's a `to_pil_image`  and `to_numpy_array` you should use here instead which handles things like any necessary type conversions, axis permutations and checking if the libraries e.g. torch are in the environment.

younesbelkada replied:
Thanks! Fixed in [17bae00](https://github.com/huggingface/transformers/pull/21400/commits/17bae00a34ded765e671beb8dfba10b2fdd32e3a)

younesbelkada committed:
fix image processors

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -292,6 +303,13 @@ def render_header(self, image: ImageInput, header: str, **kwargs):
         new_image.paste(image.resize((new_width, new_height)), (0, new_header_height))
 
+        # Convert back to the original framework if necessary
+        if return_framework is not None:
```
Thanks! Fixed in [17bae00](https://github.com/huggingface/transformers/pull/21400/commits/17bae00a34ded765e671beb8dfba10b2fdd32e3a)

younesbelkada committed:
add batched vqa support

younesbelkada committed:
fix vqa conversion

younesbelkada committed:
make fixup

amyeroberts commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1803 @@
+"""
+
+
+@add_start_docstrings(
+    "A conditional generation model with a language modeling head. Can be used for sequence generation tasks.",
+    PIX2STRUCT_START_DOCSTRING,
+)
```
This isn't correct for this model

amyeroberts reviewed (commented):
Very nice! :D 

Just a few things to get it across the finish line. Pretty much all of my comments are for the image processor. I've tried to make as many code suggestions in the review so it's easy / quick to update, in particular for docstrings. 

There's a few areas where the image processor breaks the design pattern of of image processors in the library, in particular the assumptions about inputs and outputs of the methods: 
* All the public processing methods should take and return a numpy array image. 
* The methods should also have a `data_format` argument, which defaults to `None`. 
* If `data_format` isn't specified the output data format should be the same as the input (even if it had to be change to perform the transformation in the method).
 
The reason for being strict about this is 1) so the behaviour is consistent / expected for the user 2) to avoid issues of the past when it wasn't always possible to safely use the `do_xxx` flags because of the assumptions of state and incoming image formats to methods.

younesbelkada committed:
fix conversion script

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
make fixup

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

younesbelkada committed:
add correct docstring

younesbelkada commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1803 @@
+            Flattened and padded pixel values. These values can be obtained using [`AutoImageProcessor`]. See
+            [`Pix2StructVisionImageProcessor.__call__`] for details. Check the [original
+            paper](https://arxiv.org/abs/2210.03347) (figure 5) for more details.
+
```
Thanks a lot for noticing! Fixed now!

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,407 @@
+        # [max_patches, 2 + patch_height * patch_width * image_channels]
+        result = torch.nn.functional.pad(result, [0, 0, 0, max_patches - (rows * columns)]).float()
+
+        result = to_numpy_array(result)
```
I don't think this is possible as this function returns a sequence of flattened patches. Hence the resulting shape is `max_patches, hidden_dim`. Will update the docstring

younesbelkada committed:
update docstring

younesbelkada committed:
fix module level + channel dim

younesbelkada committed:
use `make_list_of_images`

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,407 @@
+        is_vqa = self.is_vqa
+
+        if not is_batched(images):
+            images = [images]
```
Thanks for that! Fixed in f471819

younesbelkada committed:
refactor

younesbelkada commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,407 @@
+            ]
+
+        # All transformations expect numpy arrays.
+        images = [to_numpy_array(image) for image in images]
```
Fixed in [2cb4390](https://github.com/huggingface/transformers/pull/21400/commits/2cb4390248fbb3ff51b116f91247f57e8498855d)

younesbelkada committed:
correct docstring

younesbelkada committed:
fix authors

amyeroberts commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,407 @@
+        self.do_normalize = do_normalize
+        self.do_convert_rgb = do_convert_rgb
+        self.max_patches = max_patches
+        self.is_vqa = is_vqa
```
Ah, OK. That makes sense. We definitely don't want people having to pass in `is_vqa` into the processor, and I forgot that it doesn't have its own config. Fine to leave in then :)

younesbelkada replied:
Awesome thanks!

younesbelkada committed:
remove `data_format`

younesbelkada committed:
add header text test

younesbelkada commented:
Thanks a mile for the extensive review! 🚀  So from what I have got from your comment: https://github.com/huggingface/transformers/pull/21400#discussion_r1137690655 I removed the `data_format` argument
Would love to have a last round of review 💪

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add-pix2struct-clean

amyeroberts commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,419 @@
+                    - `TensorType.JAX` or `'jax'`: Return a batch of type `jax.numpy.ndarray`.
+            do_convert_rgb (`bool`, *optional*, defaults to `self.do_convert_rgb`):
+                Whether to convert the image to RGB.
+            data_format (`ChannelDimension` or `str`, *optional*, defaults to `ChannelDimension.FIRST`):
+                The channel dimension format considered by the normalization function:
+                    - `ChannelDimension.FIRST` or `'channels_first'`: The channel dimension is the first dimension.
+                    - `ChannelDimension.LAST` or `'channels_last'`: The channel dimension is the last dimension.
```
not needed anymore :) 

```suggestion
```

amyeroberts commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,419 @@
+            The patch size to use for the image. According to Pix2Struct paper and code, the patch size is 16x16.
+        max_patches (`int`, *optional*, defaults to 2048):
+            The maximum number of patches to extract from the image as per the [Pix2Struct
+            paper](https://arxiv.org/pdf/2210.03347.pdf).
```
```suggestion
            paper](https://arxiv.org/pdf/2210.03347.pdf).
        is_vqa (`bool`, *optional*, defaults to `False`):
            Whether or not the image processor is for the VQA task. If `True` and `header_text` is passed in, text is 
            rendered onto the input images.
```

amyeroberts commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,419 @@
+
+
+        Args:
+            images (`ImageInput`):
+                Image to preprocess.
+            max_patches (`int`, *optional*, defaults to `self.max_patches`):
+                Maximum number of patches to extract.
+            header_text (`Union[List[str], str]`, *optional*):
+                Text to render as a header. Only has an effect if `image_processor.is_vqa` is `True`.
+            patch_size (`dict`, *optional*, defaults to `self.patch_size`):
+                Dictionary containing the patch height and width.
+            do_normalize (`bool`, *optional*, defaults to `self.do_normalize`):
+                Whether to normalize the image.
+            return_tensors (`str` or `TensorType`, *optional*):
+                The type of tensors to return. Can be one of:
+                    - Unset: Return a list of `np.ndarray`.
+                    - `TensorType.TENSORFLOW` or `'tf'`: Return a batch of type `tf.Tensor`.
+                    - `TensorType.PYTORCH` or `'pt'`: Return a batch of type `torch.Tensor`.
+                    - `TensorType.NUMPY` or `'np'`: Return a batch of type `np.ndarray`.
+                    - `TensorType.JAX` or `'jax'`: Return a batch of type `jax.numpy.ndarray`.
+            do_convert_rgb (`bool`, *optional*, defaults to `self.do_convert_rgb`):
+                Whether to convert the image to RGB.
```
```suggestion
            images (`ImageInput`):
                Image to preprocess.
            header_text (`Union[List[str], str]`, *optional*):
                Text to render as a header. Only has an effect if `image_processor.is_vqa` is `True`.
            do_convert_rgb (`bool`, *optional*, defaults to `self.do_convert_rgb`):
                Whether to convert the image to RGB.
            do_normalize (`bool`, *optional*, defaults to `self.do_normalize`):
                Whether to normalize the image.
            max_patches (`int`, *optional*, defaults to `self.max_patches`):
                Maximum number of patches to extract.
            patch_size (`dict`, *optional*, defaults to `self.patch_size`):
                Dictionary containing the patch height and width.
            return_tensors (`str` or `TensorType`, *optional*):
                The type of tensors to return. Can be one of:
                    - Unset: Return a list of `np.ndarray`.
                    - `TensorType.TENSORFLOW` or `'tf'`: Return a batch of type `tf.Tensor`.
                    - `TensorType.PYTORCH` or `'pt'`: Return a batch of type `torch.Tensor`.
                    - `TensorType.NUMPY` or `'np'`: Return a batch of type `np.ndarray`.
                    - `TensorType.JAX` or `'jax'`: Return a batch of type `jax.numpy.ndarray`.
```

amyeroberts reviewed (approved):
Thanks for all your work adding this model ❤️  looking forward to seeing demos with it

left some tiny tiny nit suggestions but looks good to go for me!

NielsRogge commented on docs/source/en/model_doc/pix2struct.mdx:
```diff
@@ -0,0 +1,67 @@
+
+Tips:
+
+Pix2Struct has been fine tuned on a variety of tasks and datasets, ranging from natural image captioning, Visual Question Answering over different inputs (books, charts, science diagrams), captioning UI components etc. The full list can be found on the Table 1 of the paper.
+We therefore advise you to use these models for the tasks they have been fine tuned on. For instance, if you want to use Pix2Struct for UI captioning, you should use the model fine tuned on the UI dataset. If you want to use Pix2Struct for image captioning, you should use the model fine tuned on the natural images captioning dataset and so on.
```
```suggestion
Pix2Struct has been fine tuned on a variety of tasks and datasets, ranging from image captioning, visual question answering (VQA) over different inputs (books, charts, science diagrams), captioning UI components etc. The full list can be found in Table 1 of the paper.
We therefore advise you to use these models for the tasks they have been fine tuned on. For instance, if you want to use Pix2Struct for UI captioning, you should use the model fine tuned on the UI dataset. If you want to use Pix2Struct for image captioning, you should use the model fine tuned on the natural images captioning dataset and so on.
```

NielsRogge commented on src/transformers/models/pix2struct/image_processing_pix2struct.py:
```diff
@@ -0,0 +1,419 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Image processor class for PIX2STRUCT."""
```
```suggestion
"""Image processor class for Pix2Struct."""
```

NielsRogge commented on src/transformers/models/pix2struct/modeling_pix2struct.py:
```diff
@@ -0,0 +1,1807 @@
+    config_class = Pix2StructConfig
+
+    @property
+    def dummy_inputs(self):
```
Isn't this only required for TF models?

younesbelkada commented on docs/source/en/model_doc/pix2struct.mdx:
```diff
@@ -0,0 +1,67 @@
+
+This model was contributed by [ybelkada](https://huggingface.co/ybelkada).
+The original code can be found [here](https://github.com/google-research/pix2struct).
+
```
Add resources

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>
Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
make fixup

younesbelkada committed:
add checkpoints

Misterion777 commented:
hey @younesbelkda, thank you for great work, i'm not sure if this is a proper place to ask, but i haven't find a better one.
Is it possible to use pix2struct for widget captioning task, with bounding box as an additional input?
